2 New Positive Quirks!

flowercuco opened the pull request:
## About The Pull Request

I added a few quirks to a downstream that I feel fit well with tg so here they are.

First up is Poster Boy, a quirk that gives you three mood altering posters, similar to the traitor objective to hang up demoralizing posters. You spawn with a box that has one poster that will uplift the entire crews spirits and 2 that are unique to your department. Captain counts for security and assistants get only neutral posters. Finally, with a crayon or spraycan, if you are any antagonist you can make your poster into one of the ones from the traitor objective.
![dreamseeker_nRy44SL9Jb](https://user-images.githubusercontent.com/116288367/214109008-6f1b4b7c-e800-4142-be6d-926a8e975973.png)
example of quirk posters
Costs 4.


Finally, the characterful Throwing Arm quirk, which lets you throw objects further (but not harder) and means you will never miss shots into the disposals bin.  
Costs 7.

previously i had a food subscription quirk here as well but i pulled it out and plan to re-add it as a separate PR in march, where it will now give you ingredients to cook a meal with occasionally.

## Why It's Good For The Game

Positive quirk variety is good and fun, I think that these positive quirks are reasonable ones that offer unique things that the current positive quirks do not.
Poster boy gives people a reason to run around and claim wall real estate for their department and hopefully can build more solidarity in departments, the hidden antag feature probably has uses but is just for styling on people.
Throwing arm offers a fun character trait that probably can have some slight uses and encourages the use of throwing weapons and tools more. Also it is good to have a way to never miss the disposals bin. It's so embarrassing.

## Changelog
:cl:
add: Poster boy and Throwing arm positive quirks.
imageadd: added posters for poster boy quirk
/:cl:

flowercuco committed:
adds new positive quirk, poster boy and its mood altering posters

flowercuco committed:
adds throwing arm quirk

flowercuco committed:
file for giant quirk

flowercuco committed:
adds new signal that fires whenever quirks are added

flowercuco committed:
adds food order subscriber quirk

flowercuco committed:
makes timer not testing timer

ZephyrTFA commented on code/__DEFINES/traits.dm:
```diff
@@ -630,6 +630,9 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define TRAIT_ANXIOUS "anxious"
 #define TRAIT_SMOKER "smoker"
+#define TRAIT_POSTERBOY "poster_boy"
+#define TRAIT_NTGRUB_SUBCRIBER "ntgrub_subscriber"
```
```suggestion
#define TRAIT_NTGRUB_SUBSCRIBER "ntgrub_subscriber"
```

ZephyrTFA commented on code/datums/quirks/food_order_subscriber.dm:
```diff
@@ -0,0 +1,211 @@
+	desc = "You have subscribed to NTGRUB, for a low-low price you will receive regular deliveries of food tailored to your tastes."
+	icon = "gift"
+	value = 7
+	mob_trait = TRAIT_NTGRUB_SUBCRIBER
```
```suggestion
	mob_trait = TRAIT_NTGRUB_SUBSCRIBER
```

ZephyrTFA commented on code/datums/quirks/food_order_subscriber.dm:
```diff
@@ -0,0 +1,211 @@
+	mob_trait = TRAIT_NTGRUB_SUBCRIBER
+	gain_text = "<span class='danger'>You have subscribed to NTGRUB!</span>"
+	lose_text = "<span class='notice'>You feel as if your subscription to NTGRUB has been cancelled.</span>"
+	medical_record_text = "Patient is susceptible to GREAT DEALS! SUBSCRIBE TO NTGRUB NOW."
```
👎

flowercuco replied:
fair enough, it is a bit too silly

ZephyrTFA commented on code/datums/quirks/positive_quirks.dm:
```diff
@@ -207,6 +207,40 @@
+	var/department = posterboy.assigned_role.paycheck_department
+	if(department == ACCOUNT_CIV) //if you are not part of a department you instead get 3 neutral posters
+		for(var/i in 1 to 2)
+			new /obj/item/poster/quirk/crew/random/(src)
```
```suggestion
			new /obj/item/poster/quirk/crew/random(src)
```

ZephyrTFA commented on code/datums/quirks/positive_quirks.dm:
```diff
@@ -207,6 +207,40 @@
+		return
+	for(var/potential_poster in subtypesof(/obj/item/poster/quirk))
+		var/obj/item/poster/quirk/new_dept_poster = new potential_poster()
+		if(new_dept_poster.quirk_poster_department != department)
+			qdel(new_dept_poster)
+			continue
```
instead of spawning and then qdel'ing them. Check for the department being correct before spawning

ZephyrTFA commented on code/game/objects/effects/poster_motivational.dm:
```diff
@@ -0,0 +1,292 @@
+		balloon_alert(user, "interrupted!")
+		return
+	var/obj/item/poster/traitor/quirkspawn = new(get_turf(src))
+	moveToNullspace()
```
don't call this, qdeling already does this

flowercuco committed:
spelling error and stops being cheeky in medical records

flowercuco committed:
only spawns correct posters instead of spawning and deleting posters that do not match

flowercuco committed:
Merge branch 'master' into just-some-positivty

Elli-Skala commented:
Concerned the last quick will go against basketball balance in #72459

flowercuco commented:
> Concerned the last quick will go against basketball balance in #72459

do ghosts spawning into a minigame have quirks?

Fikou reviewed (changes requested):
balance of balling must be preserved (throwing arm shouldnt give a bonus to basketball hoop chance)

Fikou commented on code/datums/quirks/positive_quirks.dm:
```diff
@@ -265,6 +298,17 @@
+	name = "Throwing Arm"
+	desc = "Your arms have a lot of heft to them! Objects that you throw just always seem to fly further than everyone elses, and you never miss a toss."
+	icon = "baseball"
+	value = 2
```
extra throw range actually has a ton of value, 2 is way too low. something like a 6-8 would be better

the-orange-cow commented:
Completely unrelated, but I completely subsist off gatfruit cherry bombs and roburgers, I hope NT GRUB can accommodate my refined palette.

optimumtact commented:
I'm not sure about the NT GRUB one, seems unfair to the chef.

I'll have to think about it a bit.

MrDoomBringer commented:
Change NT GRUB to be sent through the cargo mail system, other than that good PR.

flowercuco committed:
preserves b-ball balance

davethwave commented:
To have it go through cargo would be a pretty decent way to balance this and a way to give cargo something to do occasionally.

YakumoChen commented:
Throwing stuff farther is a pretty big balance change
Would prefer its restricted to just accuracy for bins

YakumoChen commented:
Being able to more or less avoid all the management of hunger also seems pretty strong

Mothblocks commented:
@optimumtact did you decide the ingredients idea was fine?

timothymtorres commented:
> balance of balling must be preserved (throwing arm shouldnt give a bonus to basketball hoop chance)

Thank you sir!  

Other than that, everything else looks amazing! +1

SomeoneYouProbablyKnow commented:
Poster boy is just epic. Powergames will powergame it but the whole crew will benefit, that's neat.

Throwing arm is funny and based. If you want a use for it you can potentially make it increase thrown damage up to a set value like 8 ish. So anything less than 8 would go up to 8 and anything above 8 would stay as it is. 
RANDOM BULLSHIT GO

The Grub one is weird... All quirks fuck around with mood but this one goes directly through hunger. The main reason for gaming mood is the speed boost (or the lack of a speed reduction, whatever) Maybe this food should have some inherent quality that makes your combat worse? Screen shake or slowdown idk. But essentially bypassing hunger and mood all-together is very strong for 7 (I mean plasmamemes do that but shhh)

Alternatively making it unreasonable costly is also  an option. NTGrubs only serves the severely disadvantaged. Make it something like 14 and no one will be able to get it without taking on some actually negative quirks and not manageable ones heirloom or smoker

thgvr commented:
Quirks shouldn't circumvent station jobs.

BlueMemesauce commented:
> The Grub one is weird... All quirks fuck around with mood but this one goes directly through hunger. The main reason for gaming mood is the speed boost (or the lack of a speed reduction, whatever) Maybe this food should have some inherent quality that makes your combat worse? Screen shake or slowdown idk. But essentially bypassing hunger and mood all-together is very strong for 7 (I mean plasmamemes do that but shhh)

Also consider that food provides minor healing, so you're getting free healing too

flowercuco commented:
what i think is going to happen is that i will wait for feedback from oranges on the idea of this quirk and depending on the answer either make it into something less instant and more expensive or scrap it and convert it into an unrelated new service and cargo feature. 

im glad ppl like poster boy

ArthuroPep commented:
NT Grub sounds fun, but should definitely go through cargo or something. Maybe, to not circumvent the chef, have it be sent to the chef which needs to display it on their counter?
And throwing arm needs to not increase hoop chance, but recycle bins should be fine

Mothblocks commented:
we said in coderbus it would be fine if it gave you the ingredients instead to go through the chef but i dont know if that was just a musing or not

optimumtact commented:
Sorry I have had gastro the past few days so I haven't been able to follow up, I think either sending the ingredients, or making it sent through cargo would be fine options, so it's not quite a straight upgrade.

Nebulacrity commented:
I agree with the others - NT Grub should go through cargo.

Though maybe it can also be displayed on some sort of interface for Chefs to see beforehand? They can complete it themselves ahead of time for a cash reward.

YakumoChen commented:
Of the opinion the throwing quirk should be scrapped
Throw bonuses are a combat upgrade and others don't like the b-balling, leaving just disposal bin bonus, although that by itself still feels like a 3 point upgrade

flowercuco committed:
removes food delivery quirk to become a seperate pull request later

flowercuco committed:
forgot to remove the trait lol

flowercuco commented:
thank you for the review for the NTGRUB quirk, i am going to make it a separate PR when i have done the suggested changes. i dragged my feet on it for a few days bc i thought id have the energy to do it now but i dont think i will for a bit. ah well!

i've updated the op and removed it from this PR, which is now just for throwing arm and poster boy

flowercuco committed:
Merge branch 'master' into just-some-positivty

san7890 commented on code/__DEFINES/traits.dm:
```diff
@@ -630,6 +630,8 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define TRAIT_ANXIOUS "anxious"
 #define TRAIT_SMOKER "smoker"
+#define TRAIT_POSTERBOY "poster_boy"
+#define TRAIT_THROWINGARM "throwingarm"
```
```suggestion
#define TRAIT_THROWINGARM "throwing_arm"
```

snake case is swag

flowercuco replied:
this ones funny because you see me do it the correct way for the other quirk

san7890 commented on code/datums/quirks/positive_quirks.dm:
```diff
@@ -265,6 +298,17 @@
+	icon = "baseball"
+	value = 7
+	mob_trait = TRAIT_THROWINGARM
+	gain_text = "<span class='notice'>Your arms are full of energy!</span>"
+	lose_text = "<span class='danger'>Your arms ache a bit.</span>"
```
```suggestion
	gain_text = span_notice("Your arms are full of energy!")
	lose_text = span_danger("Your arms ache a bit.")
```

Using the span macros look better.

flowercuco replied:
you think its worth to do a separate pr that makes all of the other quirk notices like this into macros?

Mothblocks replied:
yes

san7890 commented on code/game/objects/effects/poster_motivational.dm:
```diff
@@ -0,0 +1,291 @@
+ * - Applies a mood buff to people who are in visible range of the item.
+ */
+/datum/proximity_monitor/advanced/quirk_posters
+	var/department
```
DMDoc what this does, not immediately obvious.

san7890 commented on code/game/objects/effects/poster_motivational.dm:
```diff
@@ -0,0 +1,291 @@
+		on_seen(examiner)
+
+/datum/proximity_monitor/advanced/quirk_posters/proc/on_seen(mob/living/viewer)
+	if (!viewer.mind)
+		return
+	if (!viewer.mob_mood)
+		return
+	if (viewer.stat != CONSCIOUS)
+		return
+	if(viewer.is_blind())
+		return
```
```suggestion
	if (!viewer.mind || !viewer.mob_mood || (viewer.stat != CONSCIOUS) || viewer.is_blind())
		return
```

Combine the four easy checks into one short-circuited line. Would be nice if you could also fold in the other two below this, but might get too long-winded for one line.

san7890 commented on code/game/objects/effects/poster_motivational.dm:
```diff
@@ -0,0 +1,291 @@
+
+/// You can use any spraypaint can on a quirk poster to turn it into a contraband poster from the traitor objective
+/obj/item/poster/quirk/attackby(obj/item/postertool, mob/user, params)
+	if(!is_special_character(user))
+		return ..()
+	if(!HAS_TRAIT(user, TRAIT_POSTERBOY))
+		return ..()
+	if(!istype(postertool, /obj/item/toy/crayon))
+		return ..()
```
```suggestion
	if(!is_special_character(user) || !HAS_TRAIT(user, TRAIT_POSTERBOY) || !istype(postertool, /obj/item/toy/crayon))
		return ..()
```

Short-circuit for brevity. https://www.byond.com/docs/ref/#/operator/||

san7890 commented on code/game/objects/effects/poster_motivational.dm:
```diff
@@ -0,0 +1,291 @@
+/// Screentip for the above
+
+/obj/item/poster/quirk/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+	if(!is_special_character(user))
+		return NONE
+	if(!HAS_TRAIT(user, TRAIT_POSTERBOY))
+		return NONE
+	if(!istype(held_item, /obj/item/toy/crayon))
+		return NONE
```
```suggestion
	if(!is_special_character(user) || !HAS_TRAIT(user, TRAIT_POSTERBOY) || !istype(held_item, /obj/item/toy/crayon))
		return NONE
```

san7890 commented on code/game/objects/effects/poster_motivational.dm:
```diff
@@ -0,0 +1,291 @@
+	name = "placeholder quirk poster"
+	desc = "Uh oh! You shouldn't have this!"
+	icon_state = "rolled_poster_legit"
+	var/quirk_poster_department = NONE
```
DMDoc this. https://github.com/tgstation/tgstation/blob/master/.github/guides/AUTODOC.md

san7890 commented on code/game/objects/effects/poster_motivational.dm:
```diff
@@ -0,0 +1,291 @@
+	var/datum/proximity_monitor/advanced/quirk_posters/mood_buff
+
+/obj/structure/sign/poster/quirk/on_placed_poster(mob/user)
+	mood_buff = new(src, 7, TRUE, quirk_poster_department)
```
Name your args. https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#prefer-named-arguments-when-the-meaning-is-not-obvious

san7890 commented on code/game/objects/effects/poster_motivational.dm:
```diff
@@ -0,0 +1,291 @@
+	RegisterSignal(host, COMSIG_PARENT_EXAMINE, PROC_REF(on_examine))
+
+/datum/proximity_monitor/advanced/quirk_posters/field_turf_crossed(atom/movable/crossed, turf/location)
+	if (!isliving(crossed))
+		return
+	if (!can_see(crossed, host, current_range))
+		return
```
```suggestion
	if (!isliving(crossed) || !can_see(crossed, host, current_range))
		return
```

san7890 commented on code/datums/quirks/positive_quirks.dm:
```diff
@@ -207,6 +207,39 @@
+		for(var/i in 1 to 2)
+			new /obj/item/poster/quirk/crew/random(src)
+		return
+	for(var/potential_poster in subtypesof(/obj/item/poster/quirk))
+		var/obj/item/poster/quirk/new_dept_poster = potential_poster
+		if(initial(new_dept_poster.quirk_poster_department) != department)
+			continue
+		new potential_poster(src)
```
```suggestion
	for(var/obj/item/poster/quirk/potential_poster as anything in subtypesof(/obj/item/poster/quirk))
		if(initial(new_dept_poster.quirk_poster_department) != department)
			continue
		new potential_poster(src)
```

This should work a bit better, although my memory is fuzzy on this type of typecasting. LMK if it breaks.

flowercuco replied:
works fine!

san7890 reviewed (changes requested):
Buncha minor stuff that should make the code look cleaner and work better. Probably missing something but I'll remember it eventually. Feel free to mark "ready for review" when everything's been addressed.

san7890 commented on code/game/objects/effects/poster_motivational.dm:
```diff
@@ -0,0 +1,291 @@
+	QDEL_NULL(mood_buff)
+	return ..()
+
+/obj/structure/sign/poster/quirk/apply_holiday()
+	var/obj/structure/sign/poster/traitor/holi_data = /obj/structure/sign/poster/quirk/festive
+	name = initial(holi_data.name)
+	desc = initial(holi_data.desc)
+	icon_state = initial(holi_data.icon_state)
```
```suggestion
/obj/structure/sign/poster/quirk/apply_holiday()
	var/obj/structure/sign/poster/traitor/holiday_data = /obj/structure/sign/poster/quirk/festive
	name = initial(holiday_data.name)
	desc = initial(holiday_data.desc)
	icon_state = initial(holiday_data.icon_state)
```

no need to abbreviate, four letters is nothing.

flowercuco replied:
btw the other holiday poster stuff (in demotivational posters and the regular poster files) use this format bc thats what the original holiday poster pr set it to

san7890 replied:
Ah, I see. We just wanna stray away from the pattern in newly added code, we can fix the old weird stuff later

san7890 commented:
btw for the future it would be appreciated if you could do PRs one-quirk-at-a-time since it was confusing for me to have to switch gears between poster code and baseball code. it's fine for now since this was originally three quirks so i get wanting to do the "DLC" stuff in one go, but it just doesn't flow as easily. good work so far though.

flowercuco committed:
Merge branch 'master' into just-some-positivty

flowercuco committed:
Merge branch 'master' into just-some-positivty

flowercuco committed:
syntax stuff

flowercuco committed:
missed variable thing

flowercuco committed:
documenting variables i neglected to document

flowercuco committed:
condenses an amount of checks into combined OR checks

flowercuco committed:
small changes and naming of arguments

flowercuco commented:
> btw for the future it would be appreciated if you could do PRs one-quirk-at-a-time since it was confusing for me to have to switch gears between poster code and baseball code. it's fine for now since this was originally three quirks so i get wanting to do the "DLC" stuff in one go, but it just doesn't flow as easily. good work so far though.

yeah sorry about that, ill keep this in mind if i ever try to do some similar stuff like this again. but most of my ideas for things in the future (right now!) are not as wide as this one I think haha

san7890 reviewed (approved):
Looks alright to me, do want @tgstation/sprite-team to look over the posters you add in this PR (and they are welcome to merge if deemed suitable).
